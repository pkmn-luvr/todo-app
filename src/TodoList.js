import React, { useState } from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';
import './TodoList.css';

function TodoList() {
  const [todos, setTodos] = useState([]);

  const addTodo = (newTodo) => {
    setTodos([...todos, {...newTodo, isCompleted: false}]); // FURTHER STUDY #2 - Included isCompleted property
  };

  const removeTodo = (id) => {
    setTodos(todos.filter(todo => todo.id !== id));
  };

  // FURTHER STUDY - Add function that updates task text in its state & passes it down to each Todo component
  const updateTodo = (id, updatedTask) => {
    setTodos(todos.map(todo => todo.id === id ? { ...todo, task: updatedTask } : todo));
  };

  // FURTHER STUDY - Strikethrough upon completion
  const toggleComplete = (id) => {
    setTodos(todos.map(todo => todo.id === id ? { ...todo, isCompleted: !todo.isCompleted } : todo));
  };

  return (
    <div>
      <NewTodoForm addTodo={addTodo} />
      {todos.map(todo => (
        <Todo
        key={todo.id}
        id={todo.id}
        task={todo.task}
        removeTodo={removeTodo}
        updateTodo={updateTodo}
        toggleComplete={toggleComplete} 
        isCompleted={todo.isCompleted} 
      />
      ))}
    </div>
  );
}


export default TodoList;
