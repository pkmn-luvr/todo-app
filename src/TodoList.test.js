import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

describe('TodoList Component Tests', () => {
  // Smoke
    test('renders without crashing', () => {
    render(<TodoList />);
  });

  // Snapshot
  test('matches snapshot', () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });

  // Business Logic Testing  - Can user add an item to todo list? 
  test('allows users to add and remove a todo', () => {
    const { getByLabelText, getByText, queryByText } = render(<TodoList />);
    const input = getByLabelText('Add a new task:');
    fireEvent.change(input, { target: { value: 'Test Todo' } });
    fireEvent.click(getByText('Add Todo'));

    // Checks if new todo is added
    expect(queryByText('Test Todo')).toBeInTheDocument();

    // Checks if todo remove button appears
    fireEvent.click(getByText('X'));
    expect(queryByText('Test Todo')).not.toBeInTheDocument();
  });
});
