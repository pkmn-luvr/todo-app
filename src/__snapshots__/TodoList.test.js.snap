// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList Component Tests matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="new-todo"
      >
        Add a new task:
      </label>
      <input
        id="new-todo"
        placeholder="Add a new task"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
</DocumentFragment>
`;
